regcomp(_internal)?.c: Avoid some casts in prints

khwilliamson opened the pull request:
We have modifiers like %z and %t to portably print items of the given type.  This converts to use those in these files, removing casts that were previously used.


* This set of changes does not require a perldelta entry.

khwilliamson committed:
regcomp(_internal).c: Avoid some casts in prints

We have modifiers like %z and %t to portably print items of the given
type.  This converts to use those in these files, removing casts that
were previously used.

leonerd commented on regcomp.c:
```diff
@@ -13147,7 +13145,7 @@ S_regtail_study(pTHX_ RExC_state_t *pRExC_state, regnode_offset p,
-                      "~ attach to %s (%" IVdf ") offset to %" IVdf "\n",
+                      "~ attach to %s (%zd) offset to %td\n",
                       SvPV_nolen_const(RExC_mysv),
                       (IV)val,
                       (IV)(val - scan)
```
You left the `(IV)` casts in; do you need to remove them for `%zd`?

khwilliamson replied:
Yes, I should have removed the casts; now done

tonycoz commented on regcomp.c:
```diff
@@ -4384,12 +4382,12 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp, U32 depth)
+            Perl_re_printf( aTHX_  "~ tying lastbr %s (%" IVdf ") to ender %s (%" IVdf ") offset %td\n",
                           SvPV_nolen_const(RExC_mysv1),
                           (IV)lastbr,
                           SvPV_nolen_const(RExC_mysv2),
                           (IV)ender,
-                          (IV)(ender - lastbr)
+                          (ender - lastbr)
```
I'm not sure if you're distinguishing `ptrdiff_t` and `ssize_t` here, or if you're using the assumption that `regnode_offset` is `SSize_t` (which it is, but I'm not sure you intend that to be visible.)

`ender` and `lastbr` are `regnode_offset`s which are `SSize_t` but you've retained the casts and `IVdf` for each of those.

I'd expect subtracting `lastbr` from `ender` to also be a `SSize_t` (%zd) but you've used `%td`.

tonycoz replied:
Similar cases below, let me know if you want me to point each of them out.

khwilliamson replied:
Thanks.  I don't know what I was thinking. This commit is 3 years old.

tonycoz commented on regcomp.c:
```diff
@@ -16329,7 +16326,7 @@ S_handle_names_wildcard(pTHX_ const char * wname, /* wildcard name to match */
      * lib/unicore/Name.pm.  'code_point' in the name is expressed in hex. */
     for (i = 0; i <= av_top_index((AV *) algorithmic_names); i++) {
-        IV j;
+        SSize_t j;
```
If I understand the code `low` and `high` represent a range of code points, and so hence `j` is also a code point.

What I'm not sure about is what controls the Perl extended valid range of code points, if it's limited to non-negative `IV` values this change is incorrect, since IV/UV can be larger than `SSize_t`.

It might be that the range here is always within the base unicode range, but I'm not sure we limit custom properties to that (or if those apply here.)

khwilliamson replied:
You're right

khwilliamson committed:
regcomp.c: Code points can be UV; not just IV

This converts from using IV to UV.  Code points are currently restricted
to 0..IV_MAX, but the intention is to eventually use values outside that
range for internal purposes.

This avoids any potential issues with printing an IV in hex format

tonycoz commented:
From the second commit message:

> but the intern is to use values outside that range for internal purposes.

should that be "but the intention ..."?
